### PR TITLE
Jonmv/separate clustercontroller cluster

### PR DESCRIPTION
**Pull request by jonmv**
@hmusum please review. 

I need some unit tests, and an actual check of whether an app is "in sync", using the `/cluster/v2` API on the CC. 
I'm also going to remove support for the `<controllers>` element under `<content>`, which is unused.

### CODE REVIEW

File: configserver/src/main/java/com/yahoo/vespa/config/server/session/SessionPreparer.java [hunk 1]
```diff
@@ -332,6 +338,8 @@ private void writeStateToZooKeeper(SessionZooKeeperClient zooKeeperClient,
             zooKeeperClient.writeDockerImageRepository(dockerImageRepository);
             zooKeeperClient.writeAthenzDomain(athenzDomain);
             zooKeeperClient.writeQuota(quota);
+            if (dedicatedClusterControllerCluster)
+                zooKeeperClient.writeDedicatedClusterControllerCluster();
```
Comment (hmusum) [inline configserver/src/main/java/com/yahoo/vespa/config/server/session/SessionPreparer.java:342]: You need to do the same in `SessionRepository.createSessionFromExisting()`, as per the comment on line 334

File: configserver/src/main/java/com/yahoo/vespa/config/server/session/SessionZooKeeperClient.java [hunk 2]
```diff
@@ -52,6 +52,7 @@
     private static final String DOCKER_IMAGE_REPOSITORY_PATH = "dockerImageRepository";
     private static final String ATHENZ_DOMAIN = "athenzDomain";
     private static final String QUOTA_PATH = "quota";
+    private static final String DEDICATED_CLUSTER_CONTROLLER_CLUSTER_PATH = "dedicatedClusterControllerCluster";
```
Comment (hmusum) [inline configserver/src/main/java/com/yahoo/vespa/config/server/session/SessionZooKeeperClient.java:55]: Consider documenting what this node is for, compared to the one on application path (and perhaps document the other one as well)

**Review by hmusum — APPROVED** [review]
Overall LGTM, two comments

**Comment by jonmv** [comment]
Thanks! Nice catch.

**Review by hmusum — APPROVED** [review]

**Comment by jonmv** [comment]
Lots of tests, and a little fix in `NodeSpecification`.

**Comment by jonmv** [comment]
4 last commits are new since you reviewed. Conflict, so had to force-push.

File: config-model/src/test/java/com/yahoo/config/model/provision/ModelProvisioningTest.java [hunk 3]
```diff
@@ -1034,6 +1074,50 @@ public void testRequiringMoreNodesThanAreAvailable() {
...
+                "  <content version='1.0' id='foo'>" +
+                "     <redundancy>1</redundancy>" +
+                "     <documents>" +
+                "       <document type='type1' mode='index'/>" +
+                "     </documents>" +
+                "     <nodes count='2' required='true'/>" +
+                "  </content>" +
+                "</services>";
+
+        int numberOfHosts = 4; // needs 2 for foo and 3 for cluster controllers.
```
Comment (hmusum) [inline config-model/src/test/java/com/yahoo/config/model/provision/ModelProvisioningTest.java:1091]: 4 is not equal to 5, so what happens here?

**Review by hmusum — APPROVED** [review]
Looks good, one minor comment. We need some testing of transitioning from cluster controllers on content nodes to dedicated nodes and vice versa, but might be easier/better to do that in an integration test

File: config-model/src/test/java/com/yahoo/config/model/provision/ModelProvisioningTest.java (hunk 3 above)
Comment (jonmv) [reply config-model/src/test/java/com/yahoo/config/model/provision/ModelProvisioningTest.java:1091]: That's why the exception is expected.

**Comment by jonmv** [comment]
Yes, we need that. I also think an integration test is the right way, since we're not keeping the state.

File: config-model/src/test/java/com/yahoo/config/model/provision/ModelProvisioningTest.java (hunk 3 above)
Comment (hmusum) [reply config-model/src/test/java/com/yahoo/config/model/provision/ModelProvisioningTest.java:1091]: D'oh 🤦

**Pull request merged**